boxes: Disable cycling using arrow keys if recipient is invalid.

zee-bit opened the pull request:
This PR disables cycling from the recipient box to any other box inside the `WriteBox` view using arrow keys, if the recipient name is invalid. This is very similar to `CYCLE_COMPOSE_FOCUS` using <kbd>Tab</kbd> key. Tests have also been updated and amended.

This PR fixes #779.

zee-bit committed:
boxes: Disable cycling using arrow keys if recipient is invalid.

This commit disables cycling from the recipient box (Stream/To) using
arrow keys if recipient is invalid, similar to Tab. Cycling through the
MessageBox now looks for arrow key events and checks for recipient
name's validity when losing focus from the recipient box. This should
prevent sending of messages to incorrect recipients.

Currently, only RIGHT and DOWN keypress is monitored since these are the
only keys that can cycle focus from the recipients box to some other box,
according to our current MessageBox layout.

Tests amended.

Fixes #779.

neiljp commented:
@zee-bit Just some quick feedback:
* is there a lot of duplication here? Could we reuse parts to make this a lot simpler?
* the tests look large and complex too
* is there some way we can trigger on 'losing focus', to avoid having all these cases? (that might cover the mouse as well, as a minor side-effect)

preetmishra reviewed (commented):
@zee-bit Hey, thanks for working on this. I pulled your changes locally and it seems to work well functionally. :+1:

However, echoing what @neiljp mentioned, we could benefit from a general approach if feasible. In any case, we would want to reduce the apparent redundancy.

zee-bit commented:
@neiljp Regarding your feedback:
- I have made two functions(`footer_notify_invalid_stream` & `footer_notify_invalid_emails`), they have somewhat increased code reusability. The main logic, I think, could not be simplified further if I am to proceed with the current implementation.
- The tests are very much related to that of `CYCLE_COMPOSE_FOCUS` as it is derived from there, but there were few places with unnecessary conditions and checks that I have removed.
- I am not quite sure if there is any other way to trigger on 'loosing focus'(without all those conditions). If you or @preetmishra have some other implementation in mind, I would love to learn :). Re mouse event, I can try extending this for mouse events as well, if this implementation looks fine.

zee-bit commented:
@zulipbot add "PR needs review"

neiljp commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -491,6 +504,48 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
+                    stream_name = (header[self.FOCUS_HEADER_BOX_STREAM]
+                                   .edit_text)
+                    stream_name_len = len(stream_name)
+                    if stream_name_len == cursor_pos:
+                        if not self.model.is_valid_stream(stream_name):
+                            self.footer_notify_invalid_stream()
+                            return key
```
This likely needs to handle the same setting of recipient ids and stream id as in the other block of very similar code - which is a good reason why this would be handled well with a refactor to handle the different keys together.

Similarly the similar block below, and another for PMs?

neiljp reviewed (commented):
@zee-bit This needs rebasing to work around conflicts and potentially update for the latest improvements in the PM UI/data, but the implementation also demonstrates both the duplication and what look like issues due to not duplicating extra code. This could be refactored variously, but we could also "up" from the message content, which might affect how the refactoring could go.

neiljp commented:
I should say that it does seem to work OK :+1: I've not found a signal emitted within urwid that would solve this for us as it stands.

zee-bit commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -491,6 +504,48 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
                 elif message != saved_draft:
                     self.view.controller.save_draft_confirmation_popup(message)
+
+        # Since, there are two keys mapped to GO_RIGHT
+        # i.e. ['right', 'l'], we dont need 'l' key to satisfy
+        # the condition: is_command_key("GO_RIGHT", key).
+        # FIXME: Refactor this condition?
+        elif key == 'right':
```
Yes.
The `l` key would otherwise, instead of getting typed as a letter in `header_box`, behave exactly the same as `right` keypress(i.e. cycle focus).

neiljp replied:
Great; can we make this behave correctly if other keys are used instead of those currently assigned? I'd prefer not to hard-code the assumption of `['right', 'l']` in the code.

zee-bit committed:
refactor: boxes: Extract footer notification for invalid stream/email.

This commit refactors code for footer notification in case of invalid
recipients (stream/email) in the WriteBox to separate helper function.
This prevents duplication of code and allows reusability.

This function can now be called with a custom error message that will
be displayed in the footer after appending it with a common typeahead
suggestion.

Tests amended.

zee-bit commented:
@zulipbot add "PR needs review"
@zulipbot remove "PR awaiting update"

prah23 commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -572,6 +586,35 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
+                            return key
+            elif key == 'down':
+                if self.focus_position == self.FOCUS_HEADER_BOX_RECIPIENT:
+                    self.update_recipient_emails(self.to_write_box)
+                    invalid_emails = self.model.get_invalid_recipient_emails(
+                        self.recipient_emails
+                    )
+                    if invalid_emails:
+                        self.footer_notify_invalid_emails(invalid_emails)
+                        return key
```
This part of the code seems the same as the section validating on cycling out using `tab`, can you perhaps add a helper within `WriteBox` to do this?

zee-bit replied:
Sure, I was thinking of adding a helper but then I hesitated, I don't know why. :)

zee-bit replied:
It will be difficult to return the `key`(incase of invalid email) from inside a helper function?

prah23 replied:
You could move the part above the `invalid_emails` to a helper (would be a prep for), but I guess it wouldn't be very necessary here since it's just a direct call to the model's method. I'll add the helper in #937 directly then.

prah23 commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -490,6 +491,19 @@ def autocomplete_emojis(self, text: str, prefix_string: str
 
+    def footer_notify_invalid_stream(self) -> None:
+        invalid_stream_error = (
+            'Invalid stream name. Use {} or {} to autocomplete.'.format(
+                primary_key_for_command('AUTOCOMPLETE'),
+                primary_key_for_command('AUTOCOMPLETE_REVERSE'))
```
It might be more vivid to style the commands in `'code'`, similar to the way we show other commands in hints in the footer.

prah23 reviewed (changes requested):
@zee-bit Thanks for this!
This works as expected and restricts the user from bypassing the validation quite well. 👍 
I've dropped a couple of comments inline, as some of these changes would also benefit #937.

neiljp reviewed (commented):
@zee-bit Extracting the helper(s) for the footer simplifies things to some degree, but the overlap between tab and right/down is still very similar. I'm currently fairly open to this PR since you have a test. However, the code is highly duplicated and I'd much prefer a solution which combined the two branches if we are able, to avoid that. We can always refactor later, but then we accumulate "technical debt".

zee-bit commented:
@zulipbot add "PR needs review"
@zulipbot remove "PR awaiting update"

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -490,6 +490,16 @@ def autocomplete_emojis(self, text: str, prefix_string: str
+            ('code', primary_key_for_command('AUTOCOMPLETE_REVERSE')),
+            ' to autocomplete.'
+        ]
+        self.view.set_footer_text(footer_message, 3)
```
The method is calling an existing method, `set_footer_text`, with a particular error message. This could lead us to a state where we have multiple methods that essentially do the same thing with different arguments for different _error_ types.

Another way could be to just have an `ERROR_TEMPLATES` data structure that could be used to pass an appropriate error message to the existing `set_footer_text` method.

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -500,6 +501,16 @@ def footer_notify_invalid_recipient(self, error_msg: str) -> None:
+        stream_name_len = len(stream_header.edit_text)
+        cursor_pos = stream_header.edit_pos
+        if (is_command_key('CYCLE_COMPOSE_FOCUS', key)
+                or key in ['right', 'down']):
```
As these command keys are configurable from `keys.py`, we shouldn't hardcode them here.

zee-bit replied:
Actually, the command that binds these keys(`right` and `down`) offers a secondary letter-key(i.e. `i` and `j` respectively). Hence, using the commands instead of keys would prevent the letter keys from getting typed in the write-box.

preetmishra replied:
I understand why you've used them. However, you can see that if the keys change from `keys.py`, it is not going to where. For example, we could have another method for figuring out whether the command keys have letters and take it further from there. In any case, hardcoding shouldn't be our last resort.

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -500,6 +501,16 @@ def footer_notify_invalid_recipient(self, error_msg: str) -> None:
         ]
         self.view.set_footer_text(footer_message, 3)
 
+    def loosing_focus_from_stream_box(self, key: str) -> bool:
```
There a typo in the method name and in the commit description.

zee-bit replied:
I can see a typo in the commit description - "diables" in the first line?
But, I can't figure out the typo in the function name. Could you please clarify?

preetmishra replied:
Yeah, re the latter, it's lose vs loose.

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -500,6 +501,16 @@ def footer_notify_invalid_recipient(self, error_msg: str) -> None:
+        if (is_command_key('CYCLE_COMPOSE_FOCUS', key)
+                or key in ['right', 'down']):
+            if key != 'right' or stream_name_len == cursor_pos:
+                return True
```
The nested conditional could be simplified with what follows.
```python3
        if (is_command_key('CYCLE_COMPOSE_FOCUS', key) or key == "down"
              or (key == "right" and stream_name_len == cursor_pos)):
            return True
 ```
 
 While writing conditionals, expressing the intent clearly is vital. For instance, in the snippet above it is comparatively apparent that we're checking whether the key is `TAB` or `DOWN` or `RIGHT` (given that the cursor is at the end) for blur ("focus loss") event. Both approaches work, but the latter is more comprehensible.

zee-bit replied:
Yep, this is definitely more clear and comprehensible than the earlier conditional. Guess I'll have to look from different angles if the conditionals are getting complex. Thanks. :+1:

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -592,32 +604,37 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
-                        self.recipient_user_ids = user_ids
-                        self.stream_id = self.model.stream_id_from_name(
-                                                                stream_name)
+                        if not is_primary_direction_key(key):
```
In place of recurring conditionals, we could just check for it once and return early? That would remove all the following code changes.

zee-bit replied:
All of these conditionals have their corresponding `else`/`elif` clauses, so returning early might not give the expected behavior. Let me know if I misunderstood you. :)

preetmishra replied:
Right, we would have to return early from three places.

I think a cleaner way would be to provision another `elif` and check the stream validity separately.

preetmishra commented on tests/ui_tools/test_boxes.py:
```diff
@@ -709,7 +737,8 @@ def test_keypress_CYCLE_COMPOSE_FOCUS(self, write_box, tab_key,
         else:
             write_box.private_box_view()
-        size = widget_size(write_box)
+        # FIXME: Use widget_size(), instead of hard-coded size.
```
Why aren't we using `widget_size`?

zee-bit replied:
Well, I tried using `widget_size`, but for some cases, the `write-box` changes to box widget instead of flow and returns a `(maxcol, maxrow)` size instead of just `(maxcol, )`. This made the tests fail for such cases. Let me try to dig more into this and see if I can get a fix for this.

zee-bit replied:
Sorry for the late reply. I got busy with other PR's since finding the root+fix of this particular issue was taking forever. :(
I have tried to dig more on this, but it seems like it's an internal `urwid` issue that I am unable to resolve. While looking at this, I realized that the whole `write_box` is sometimes recognized as "flow widget" for all tests (that's when all the tests pass), but sometimes the `write_box` is determined as "box widget" for all tests (2 tests fail in this case).

I also found that in the latter case, the tests only fail for "DOWN" keypresses, because it can't be handled properly by `super()` i.e. urwid's `container.py`. I can fix the whole issue by not letting the "DOWN" keypress being handled by urwid and returning early after manually handling, but it seems to have other side-effects and does not look clean.

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -582,7 +593,8 @@ def keypress(self, size: urwid_Size, key: str) -> Optional[str]:
                     )
-        elif is_command_key('CYCLE_COMPOSE_FOCUS', key):
+        elif (is_command_key('CYCLE_COMPOSE_FOCUS', key)
+              or is_primary_direction_key(key)):
```
Why are capturing all the direction keys with the method? Wouldn't `GO_RIGHT` and `GO_DOWN` suffice?

zee-bit replied:
Actually, yes they should.
I may have misinterpreted [one of Neil's point from above](https://github.com/zulip/zulip-terminal/pull/864#pullrequestreview-591884868), and thought we needed to handle cycling for the "up" key as well. I'll amend the commit to use only `GO_RIGHT` and `GO_DOWN`. Then, I guess we won't be needing the second commit at all? I'll remove that too and keep a backup of this branch just in case. :)

preetmishra commented on zulipterminal/ui_tools/boxes.py:
```diff
@@ -490,6 +490,16 @@ def autocomplete_emojis(self, text: str, prefix_string: str
+            ('code', primary_key_for_command('AUTOCOMPLETE_REVERSE')),
+            ' to autocomplete.'
+        ]
+        self.view.set_footer_text(footer_message, 3)
```
The method is calling an existing method, `set_footer_text`, with a particular error message. This could lead us to a state where we have multiple methods that essentially do the same thing with different arguments for different _error_ types.

Another way could be to just have an `ERROR_TEMPLATES` data structure that could be used to pass an appropriate error message to the existing `set_footer_text` method.

zee-bit replied:
Would it make sense to define the `ERROR_TEMPLATES` in a separate module?

preetmishra reviewed (commented):
@zee-bit Thanks for your work. I pulled code locally and it seems to work well! :+1: 

While the PR is functional, we could simplify the code a bit. I have left a few in-line comments. Let me know if there's anything that needs clarification.
